### ISSUE TITLE
BUG: float64.__lt__ should return NotImplemented

### ISSUE THREAD

**Issue by oscarbenjamin** [numpy/numpy#22535]
### Describe the issue:

Here `numpy.float64.__lt__` is doing something weird that seems to take the result of `x.__gt__` and call `bool` on it. It should just return `NotImplemented` to let `x.__gt__` handle the operation.
```python
In [1]: import sympy

In [2]: import numpy

In [3]: x = sympy.symbols('x')

In [4]: xx = numpy.float64(0.1)

In [5]: x > xx
Out[5]: x > 0.1

In [6]: xx < x
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-af13ae84b73d> in <module>
----> 1 xx < x

~/current/sympy/sympy.git/sympy/core/relational.py in __bool__(self)
    509 
    510     def __bool__(self):
--> 511         raise TypeError("cannot determine truth value of Relational")
    512 
    513     def _eval_as_set(self):

```
Note that an ordinary `float` works:
```python
In [7]: float(xx) < x
Out[7]: x > 0.1

In [8]: float(xx).__lt__(x)
Out[8]: NotImplemented
```

### Reproduce the code example:

```python
import numpy as np
import sympy
x = sympy.symbols('x')
xx = np.float64(0.1)
xx < x
```

### Error message:

```shell
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/current/sympy/sympy.git/t.py in <module>
      3 x = sympy.symbols('x')
      4 xx = np.float64(0.1)
----> 5 xx < x

~/current/sympy/sympy.git/sympy/core/relational.py in __bool__(self)
    509 
    510     def __bool__(self):
--> 511         raise TypeError("cannot determine truth value of Relational")
    512 
    513     def _eval_as_set(self):

TypeError: cannot determine truth value of Relational
```

### NumPy/Python version information:

In [9]: import sys, numpy; print(numpy.__version__, sys.version)
1.23.4 3.8.9 (default, Apr  3 2021, 01:02:10) 
[GCC 5.4.0 20160609]

### Context for the issue:

This is coming from SO:
https://stackoverflow.com/questions/74326008/type-error-in-sympy-piecwise-python-cannot-determine-truth-value-of-relational/74327769#74327769

**Comment by asmeurer** [comment]
The problem is that no NumPy operations return NotImplemented. They always try to absorb the argument into the array. The NumPy dispatching is more complicated than the Python defaults and can be customized with `__array_ufunc__`. I believe SymPy can solve this by setting `__array_ufunc__ = None`. See https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_ufunc__. (side note: are the docs there wrong? They say setting it to `None` makes `__add__` raise `TypeError` but should it say that it makes `__add__` return `NotImplemented`, making `+` raise `TypeError` if there is no `__radd__` fallback?)

Maybe you could argue that the scalar float64(0.1) should work differently from the 0-d array np.array(0.1, dtype=float64). Personally I don't like NumPy scalars and am always annoyed whenever they work differently from 0-d arrays.

**Comment by seberg** [comment]
> side note: are the docs there wrong? They say setting it to None makes __add__ raise 

Seems like the first sentence should say unconditionally returns `NotImplemented`, yeah.

The scalars only work differently for sequence multiplication, because that is an annoying trap.
I would be happy if scalars would not pretend to be arrays at all.  But I think that can only work out if we can decide on a transition to stop converting 0-D arrays to scalars randomly.

**Comment by seberg** [comment]
I don't think this is a bug, there is little we can reasonably do.  The docs could use a slight update as Aaron pointed out.

**Comment by oscarbenjamin** [comment]
I don't know how to see the code that handles this in NumPy but if I understand correctly the situation is that `xx.__lt__` attempts to create a `bool`. It knows that it can't handle the type of `x` so it calls `x.__gt__` and then it tries to convert that to a `bool`. I presume this is the case because if `xx.__lt__` did the normal thing of returning `NotImplemented` then `x.__gt__` would be called but its result would be returned as is so it must be `xx.__lt__` that is calling `x.__gt__` somewhere.

If this is the case then why when it can't handle the input type does it not just return `NotImplemented` to allow the cooperative dispatch mechanism to work?

**Comment by seberg** [comment]
Ah, the problem here is that NumPy attempts to return a boolean value.  Things work out fine (through annoying black magic), but NumPy comparisons enforce a boolean result and that fails:
```
bool(0.1 < x)  # fails, no NumPy involved
```
`np.less(xx, x, dtype=object)` would even work.  Not sure whether there would be a point of relaxing this for the scalar path (making scalars more special and adding some code path unfortunately).

**Comment by oscarbenjamin** [comment]
Potentially it could be useful to have something like this in the non-scalar case as well:
```python
In [4]: np.less(np.array([1, 2]), x, dtype=object)
Out[4]: array([x > 1, x > 2], dtype=object)
```
Alternatively it might be more useful just for SymPy (and other code) to be able to control what happens here.

**Comment by jorenham** [comment]
Not sure when this was solved, but it seems to work now 🤷🏻 

```python
In [1]: import sympy as sp

In [2]: import numpy as np

In [3]: x = sp.symbols("x")

In [4]: xx = np.float64(.1)

In [5]: x > xx
Out[5]: x > 0.1

In [6]: xx < x
Out[6]: x > 0.1

In [7]: np.__version__
Out[7]: '2.2.1'

```
